### PR TITLE
StoreState Class Documentation

### PR DESCRIPTION
**Pull request by Ha02hen**
### Participants
@Ha02hen @Echo11ovo 

### Types of change

- [x] The form of associating the index of the component in the StoreState class has changed from an `int` type to a `str` type
- [x] Added a feature to read the path and query parameters from the URL and save them to local storage.
- [x] The functionality to read data from local storage and save it back to the URL.

### Current Data Flow Diagram

From the flow diagram below, we can see that entering a new URL in the browser will trigger an event. If it is determined that the ID of the triggered event is "URL", only then will the local storage change. Similarly, the process from local storage to URL follows a similar pattern. Additionally, if the state of a certain component changes, it will also trigger an event, and the triggered component's ID will be recorded. The corresponding key and value are then saved to local storage. Moreover, when the local storage changes, the value of the "timestamp" key will also change, further affecting the web state. Without these triggers and timestamps, it could potentially result in an infinite loop in the Dash callback.
23b0cc3d398e540d8a854a7367c8981

### Configuration and Setup
1. Place the `store_state.py` file into the `components` folder of your project.

2. Import the `StoreState` class in your code to utilize its functionalities:
`from components.store_state import StoreState`

3. Create an instance of StoreState and include a list data type that specifies which component IDs on the page need to be persisted. Set the storage type accordingly; it is recommended to use 'local', but 'session' can be chosen depending on specific requirements.
`store_state = StoreState(PERSIST_IDS, store_type='local')  # Set storage type to 'local' or 'session'`
Below is an example of a list:
```
PERSIST_IDS = [
    "persist_0",
    "persist_1",
    "persist_a",
    "persist_b",
    "model",
    "type"
]
```

4. Within the required component, set up the Store component using the get_store_component method:
`store_state.get_store_component('web_state')  # Configure the Store component`
`dcc.Location(id='url', refresh=False)`

5. Use the `setup_dash_callbacks` method to configure the callbacks for the app, ensuring correct connection with the Store component:
`store_state.setup_dash_callbacks(app, store_id='web_state')  # Set up callbacks`

#### Development Environment:
Dash version: 2.17.1
React version: 18.2.0
Python version: 3.12
Operating System: Windows 11

### Testing and Demonstration
When a URL is entered in the browser, the content in the path and query parameters will be saved to local storage and reflected in the webpage's state.
image

### Demo code
Try creating a file named `xxx.py` in the root directory of the project with the code provided below, and you should be able to run it successfully.

```
import dash
import dash_mantine_components as dmc
from dash import dcc
from components.store_state import StoreState

dash._dash_renderer._set_react_version('18.2.0')

# Initialize the Dash app
app = dash.Dash(__name__)

# List of component IDs to be persisted in local storage
PERSIST_IDS = [
    "persist_0",
    "persist_1",
    "persist_a",
    "persist_b",
    "model",
    "type"
]

# Create an instance of StoreState to manage local storage with specified component IDs
store_state = StoreState(PERSIST_IDS, store_type='local')

# Define the app layout using Mantine components
app.layout = dmc.MantineProvider(
    children=[
        dmc.Center(
            children=[
                dmc.Text(
                    "Store State Management Example",
                    variant="gradient",
                    gradient={"from": "red", "to": "yellow", "deg": 400},
                    style={"fontSize": 40, "marginBottom": 20},
                ),
            ]
        ),

        # Stack components to align them in the layout
        dmc.Stack(
            children=[
                # Container for input components with fluid width and specific width settings
                dmc.Container(
                    children=[
                        # Slider component with persistence enabled
                        dmc.Slider(
                            id="persist_0",
                            value=20,
                            marks=[
                                {"value": 20, "label": "20%"},
                                {"value": 50, "label": "50%"},
                                {"value": 80, "label": "80%"},
                            ],
                            mb=35,
                        ),
                        # Select component with options to choose from
                        dmc.Select(
                            placeholder="Select one",
                            id="persist_1",
                            value="ashrae-55",
                            data=[
                                {"value": "ashrae-55", "label": "ashrae-55"},
                                {"value": "en-16798", "label": "en-16798"},
                            ],
                            w=600,
                            mb=10,
                        ),
                        # TextInput components for input with persistence enabled
                        dmc.TextInput(
                            id="persist_a",
                            placeholder="Type model...",
                            value="20",
                            w=600,
                            mb=10,
                        ),
                        dmc.TextInput(
                            id="persist_b",
                            placeholder="Type model...",
                            value="20",
                            w=600,
                            mb=10,
                        ),
                        dmc.TextInput(
                            id="model",
                            placeholder="Type model...",
                            value="20",
                            w=600,
                            mb=10,
                        ),
                        dmc.TextInput(
                            id="type",
                            placeholder="Type type...",
                            value="20",
                            w=600,
                            mb=10,
                        ),
                    ],
                    fluid=True,
                    w=600,
                ),

                # Include the Store component to manage state persistence across sessions
                store_state.get_store_component('web_state'),

                # Location component to manage URL changes without refreshing the page
                dcc.Location(id='url', refresh=False),
            ],
            align="center",
            gap="xl"
        )
    ],
    withCssVariables=True,
    defaultColorScheme='light',
)

# Set up callbacks for handling state changes and URL parameters
store_state.setup_dash_callbacks(app, store_id='web_state')

# Run the Dash app
if __name__ == '__main__':
    app.run_server(debug=True)

```

### CODE REVIEW

File: components/store_state.py [hunk 1]
```diff
@@ -26,47 +28,169 @@ def load_state_from_store(store_data, key):
...
+            [Input(id, "value") for id in self.persist_ids],
+            [Input("url", "href")],
             State(store_id, "data"),
             prevent_initial_call=True,
         )
-        def save_data(input_values, store_data):
-            # Update storage if triggered by input component
+        def manage_store_data(*args):
+            # The URL is the second-to-last argument
+            url = args[-2]
```
Comment (FedericoTartarini) [inline components/store_state.py:58]: I do not like to use positional arguments, what -2 refers to? can we avoid doing this?

File: components/store_state.py [hunk 2]
```diff
@@ -26,47 +28,169 @@ def load_state_from_store(store_data, key):
             return store_data[key]
         return None
 
+    @staticmethod
+    def parse_url(url):
```
Comment (FedericoTartarini) [inline components/store_state.py:32]: add type annotations to functions

File: components/store_state.py [hunk 3]
```diff
@@ -26,47 +28,169 @@ def load_state_from_store(store_data, key):
...
+        """Parse URL and extract useful information."""
+        url = str(url)
+        parsed_url = urlparse(url)
+        query_params = parse_qs(parsed_url.query)
+        path_parts = parsed_url.path.strip("/").split("/")
+
+        return {
```
Comment (FedericoTartarini) [inline components/store_state.py:39]: return a class not a dictionary

File: components/store_state.py [hunk 4]
```diff
@@ -26,47 +28,169 @@ def load_state_from_store(store_data, key):
...
+        @app.callback(
+            Output("url", "href"),
+            [Input(id, "value") for id in self.persist_ids],
+            [Input("url", "href")],
+            State(store_id, "data"),
+            prevent_initial_call=True,
+        )
+        def update_url(*args):
+            # The URL is the second-to-last argument
+            url = args[-2]
```
Comment (FedericoTartarini) [inline components/store_state.py:144]: this is a code duplication, consider creating a function

File: components/store_state.py [hunk 5]
```diff
@@ -14,7 +16,7 @@ def get_store_component(self, store_id):
     @staticmethod
     def save_state_to_store(store_data, key, value):
         """Save data to dcc.Store."""
-        if store_data is None:
+        if store_data is None or not isinstance(store_data, dict):
             store_data = {}
         store_data[key] = value
```
Comment (FedericoTartarini) [inline components/store_state.py:21]: are the keys the input data? If so please make sure we use the same id as in the main application.

**Review by FedericoTartarini — COMMENTED** [review]
I think you have done a very good job but I am struggling to understand the code. It would be nice to meet separately and discuss the code line by line. I will accept the pull request once you address my comments. Well done and thank you for working on this.

File: components/store_state.py (hunk 3 above)
Comment (Ha02hen) [reply components/store_state.py:39]:
```
@staticmethod
    def parse_url(url):
        """Parse the URL and extract useful information, returning a ParsedURL object."""
        url = str(url)
        parsed_url = urlparse(url)
        query_params = parse_qs(parsed_url.query)
        path_parts = parsed_url.path.strip("/").split("/")

        # Return a ParsedURL object
        return ParsedURL(
            path=parsed_url.path,
            query_params=query_params,
            fragment=parsed_url.fragment,
            path_parts=path_parts
        )
```
Now the code has been changed to this

**Comment by FedericoTartarini** [comment]
@Ha02hen please address the other comments

File: components/store_state.py (hunk 4 above)
Comment (Ha02hen) [reply components/store_state.py:144]:
```
def update_url(*args):
    *id, url, store_data = args
```
Now the code has been changed to this

**Review by FedericoTartarini — APPROVED** [review]

**Pull request merged**